lib/std/zig/system/NativeTargetInfo.zig

```diff
@@ -355,37 +355,55 @@ fn detectAbiAndDynamicLinker(
         return result;
     }
 
-    const env_file = std.fs.openFileAbsoluteZ("/usr/bin/env", .{}) catch |err| switch (err) {
-        error.NoSpaceLeft => unreachable,
-        error.NameTooLong => unreachable,
-        error.PathAlreadyExists => unreachable,
-        error.SharingViolation => unreachable,
-        error.InvalidUtf8 => unreachable,
-        error.BadPathName => unreachable,
-        error.PipeBusy => unreachable,
-        error.FileLocksNotSupported => unreachable,
-        error.WouldBlock => unreachable,
-        error.FileBusy => unreachable, // opened without write permissions
-
-        error.IsDir,
-        error.NotDir,
-        error.InvalidHandle,
-        error.AccessDenied,
-        error.NoDevice,
-        error.FileNotFound,
-        error.FileTooBig,
-        error.Unexpected,
-        => return defaultAbiAndDynamicLinker(cpu, os, cross_target),
+    const elf_file = blk: {
+        // This block looks for a shebang line in /usr/bin/env,
+        // if it finds one, then instead of using /usr/bin/env as the ELF file to examine, it uses the file it references instead,
+        // doing the same logic recursively in case it finds another shebang line.
+
+        // Since /usr/bin/env is hard-coded into the shebang line of many portable scripts, it's a
+        // reasonably reliable path to start with.
+        var file_name: []const u8 = "/usr/bin/env";
+        while (true) {
+            const file = std.fs.openFileAbsolute(file_name, .{}) catch |err| switch (err) {
+                error.NoSpaceLeft => unreachable,
+                error.NameTooLong => unreachable,
+                error.PathAlreadyExists => unreachable,
+                error.SharingViolation => unreachable,
+                error.InvalidUtf8 => unreachable,
+                error.BadPathName => unreachable,
+                error.PipeBusy => unreachable,
+                error.FileLocksNotSupported => unreachable,
+                error.WouldBlock => unreachable,
+                error.FileBusy => unreachable, // opened without write permissions
+
+                error.IsDir,
+                error.NotDir,
+                error.InvalidHandle,
+                error.AccessDenied,
+                error.NoDevice,
+                error.FileNotFound,
+                error.FileTooBig,
+                error.Unexpected,
+                => return defaultAbiAndDynamicLinker(cpu, os, cross_target),
 
-        else => |e| return e,
+                else => |e| return e,
+            };
+
+            // #! (2) + 255 (max length of shebang line since Linux 5.1) + \n (1)
+            var buffer: [258]u8 = undefined;
+            const line = file.reader().readUntilDelimiter(&buffer, '\n') catch break :blk file;
+            if (mem.startsWith(u8, line, "#!")) {
+                file_name = std.mem.tokenize(u8, line[2..], " ").next().?;
```

Comment:
```suggestion
                var it = std.mem.tokenize(u8, line[2..], " ");
                file_name = it.next() orelse line[2..]; 
```
I don't think we should assert that the line contains a space.

Reply:
We don't? It works perfectly fine without spaces in file, `tokenize` returns entire line if it can't find delimiter.

Reply:
Right, I looked at the tests incorrectly, it won't panic if there are no spaces but it will panic if `line[2..].len == 0`.

Reply:
> Right, I looked at the tests incorrectly, it won't panic if there are no spaces but it will panic if `line[2..].len == 0`.

Should we fall back to default if this is an incomplete shebang?

```suggestion
                if (line[2..].len == 0) return defaultAbiAndDynamicLinker(cpu, os, cross_target);
                file_name = std.mem.tokenize(u8, line[2..], " ").next().?;
```

Reply:
That sounds reasonable.

Reply:
> That sounds reasonable.

Done!

Reply:
> I assigned the result of `tokenize` to a new variable deliberately, using it directly like this is invalid in stage2. Also why not use `orelse` when it is right there?

Sorry, done